EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc

```diff
@@ -224,6 +248,190 @@ namespace l1t {
+                                               UCTCTP7RawData_HCALFB& ctp7Data_HCALFB,
+                                               HcalTrigPrimDigiCollection* hcalTPGs) {
+      UCTCTP7RawData_HCALFB::CaloType cType = UCTCTP7RawData_HCALFB::HF;
+      for (uint32_t side = 0; side <= 1; side++) {
+        bool negativeEta = false;
+        if (side == 0)
+          negativeEta = true;
+        for (uint32_t iEta = 30; iEta <= 40; iEta++) {
+          for (uint32_t iPhi = 0; iPhi < 2; iPhi++) {
+            if (iPhi == 1 && iEta == 40)
+              iEta = 41;
+            int cPhi = 1 + lPhi * 4 + iPhi * 2;  // Calorimeter phi index: 1, 3, 5, ... 71
+            if (iEta == 41)
+              cPhi -= 2;                  // Last two HF are 3, 7, 11, ...
+            cPhi = (cPhi + 69) % 72 + 1;  // cPhi -= 2 mod 72
+            int cEta = iEta;
+            if (negativeEta)
+              cEta = -iEta;
+            // This code is fragile! Note that towerDatum is packed as is done in HcalTriggerPrimitiveSample
+            // Bottom 8-bits are ET
+            // Then feature bit
+            // Then minBias ADC count bit
+            // The remaining bits are undefined presently
+            // We use next three bits for link details, which we did not have room in EcalTriggerPrimitiveSample case
+            // We use next three bits to set the tower masking, link masking and link status information as done for Ecal
+            // To decode these custom six bits use ((EcalTriggerPrimitiveSample::raw() >> 9) & 0x77)
+            uint32_t towerDatum = ctp7Data_HCALFB.getET(cType, negativeEta, iEta, iPhi);
+            towerDatum |= ctp7Data_HCALFB.getFB(cType, negativeEta, iEta, iPhi) << 8;
+            if (ctp7Data_HCALFB.isLinkMisaligned(cType, negativeEta, iEta, iPhi))
+              towerDatum |= 0x0400;
+            if (ctp7Data_HCALFB.isLinkInError(cType, negativeEta, iEta, iPhi))
+              towerDatum |= 0x0800;
+            if (ctp7Data_HCALFB.isLinkDown(cType, negativeEta, iEta, iPhi))
+              towerDatum |= 0x1000;
+            if (ctp7Data_HCALFB.isTowerMasked(cType, negativeEta, iEta, iPhi))
+              towerDatum |= 0x2000;
+            if (ctp7Data_HCALFB.isLinkMasked(cType, negativeEta, iEta, iPhi))
+              towerDatum |= 0x4000;
+            if (ctp7Data_HCALFB.isLinkMisaligned(cType, negativeEta, iEta, iPhi) ||
+                ctp7Data_HCALFB.isLinkInError(cType, negativeEta, iEta, iPhi) ||
+                ctp7Data_HCALFB.isLinkDown(cType, negativeEta, iEta, iPhi))
+              towerDatum |= 0x8000;
+            HcalTriggerPrimitiveSample sample(towerDatum);
+            HcalTrigTowerDetId id(cEta, cPhi);
+            id.setVersion(1);  // To not process these 1x1 HF TPGs with RCT
+            HcalTriggerPrimitiveDigi tpg(id);
+            tpg.setSize(1);
+            tpg.setSample(0, sample);
+            hcalTPGs->push_back(tpg);
+          }
+        }
+      }
+    }
```

Comment:
There are multiple instances of code that looks very similar to this. Can any of this be abstracted to avoid repeated code?

Reply:
These are made explicit and separate so that all 4 unpackers (normal vs 5BX and HCALFB impl vs not) are completely independent to each other and for easier debugging with FW in future if we want to test/modify one of them at a time.